### ISSUE TITLE
generated serializer not properly

### ISSUE THREAD

**Issue by naotakat** [msgpack/msgpack-cli#156]
I'm using msgpack-cli 0.6.8.
I created test project for generating serializer in my solution.

Case 1.
I can not generate properly source code when using target type placed another project(referenced) in same solution.

Result Err below
(I'm sorry that is japanese..)
テスト メソッド UnitTest.GenerateSerializer.Generate が例外をスローしました:
System.IO.FileLoadException: ファイルまたはアセンブリ 'MsgPack, Version=0.6.0.0, Culture=neutral, PublicKeyToken=null'、またはその依存関係の 1 つが読み込めませんでした。見つかったアセンブリのマニフェスト定義はアセンブリ参照に一致しません。 (HRESULT からの例外:0x80131040)

Case 2 
Creating from DLL
like below
var assembly = Assembly.LoadFrom("../../../Libs/SharedModel.dll");
SerializerGenerator.GenerateSerializerSourceCodes(
settings, assemlby.GetType().Where(//only datacontract));

Result
Created! But code for polymorphism(MessagePackKnownCollectionItemType) not generated .

Case 3 
when used type copied from another project to under same project, it worked with test code.

Let me know.. plz. is it bug? or should I use latest 0.7.0 version?
I prefer to use Case 2.

**Comment by yfakariya** [comment]
Thank you for reporting!

First, 0.7 is still alpha, so 0.6.8 is preferred. 
I cannot determine your issue is true bug or misuse (but I know it suggests lacking samples/documentations...)

Case1: 
Would you tell me full stack trace for it? It looks very standard case which similar to the unittest of Msgpack for CLI itself. In addition, would you tell me what platform did you use? (Desktop .NET 4.5.2, UWP 10, .NET Core RC1-xxxxx, etc)
Case2: Make sure you qualify your member(s) with `[MessagePackKnownType]` attributes. Or, if you can do, please show your type which should be polymorphic.

**Comment by naotakat** [comment]
you can test Case1, Case2 from below link that I created solution for testing.
that includes 2 test methods for case1, 2.

https://drive.google.com/file/d/0B1HB2Z2zOgpxQ21VR1RLdmRjVDA/view?usp=sharing

**Comment by yfakariya** [comment]
Sorry for delay.

Case1:
I found that your `SharedModel` project refers ../Libs/MsgPack.dll which was built for .NET 2.0, but your test project refers NuGet package and actually point the dll for .NET 4.5. So, there is a assembly mismatch between your `SharedModel` and `UnitTest` as the exception message said.

Case2:
I checked `SharedModel_A_B_TestC_ProblemClassSerializer.cs` and it has polymorphism support codes from line 50 to line 54. Do I misread your problem?

**Comment by naotakat** [comment]
Thank you for replying.

Case1:
I need to use SharedModel on Server(.net 4.6.1) and Client(Unity 5.3 ios, android ).
So I created ShraredModel with target framework Unity 3.5 full base class libraries and unity-full msgpack dll.
and For generating serializer , Created Project with .net 4.5 with nuget package and msgpack 4.5 dll. When I tested unity-full dll in test project. SerializerGenerator not found. 
If I would like to use both platform unity, server, I cannot use .net 4.5. how should I use it and which dll does I use?
Sorry that my english sucks..

Case2:
Serializer created from dll seems not to support polymorphism below.
`
public SharedModel_A_B_TestC_ProblemClassSerializer(MsgPack.Serialization.SerializationContext context) : 
                base(context) {
            this._serializer0 = context.GetSerializer<SharedModel.A.Error>(MsgPack.Serialization.EnumMessagePackSerializerHelpers.DetermineEnumSerializationMethod(context, typeof(SharedModel.A.Error), MsgPack.Serialization.EnumMemberSerializationMethod.Default));
            this._serializer1 = context.GetSerializer<SharedModel.A2.B2.DataType>(MsgPack.Serialization.EnumMessagePackSerializerHelpers.DetermineEnumSerializationMethod(context, typeof(SharedModel.A2.B2.DataType), MsgPack.Serialization.EnumMemberSerializationMethod.Default));
            MsgPack.Serialization.PolymorphismSchema schema0 = default(MsgPack.Serialization.PolymorphismSchema);
            schema0 = null;
            this._serializer2 = context.GetSerializer<System.Collections.Generic.List<SharedModel.A2.B2.DataBase>>(schema0);
            MsgPack.Serialization.PolymorphismSchema schema1 = default(MsgPack.Serialization.PolymorphismSchema);
            schema1 = null;
            this._serializer3 = context.GetSerializer<long>(schema1);
            MsgPack.Serialization.PolymorphismSchema schema2 = default(MsgPack.Serialization.PolymorphismSchema);
            schema2 = null;
            this._serializer4 = context.GetSerializer<string>(schema2);
            MsgPack.Serialization.PolymorphismSchema schema3 = default(MsgPack.Serialization.PolymorphismSchema);
            schema3 = null;
            this._serializer5 = context.GetSerializer<System.Nullable<SharedModel.A.Error>>(schema3);
            MsgPack.Serialization.PolymorphismSchema schema4 = default(MsgPack.Serialization.PolymorphismSchema);
            schema4 = null;
            this._serializer6 = context.GetSerializer<System.Nullable<SharedModel.A2.B2.DataType>>(schema4);
        }
`
it need to add like below code in constructor.
`
        public SharedModel_A_B_TestC_ProblemClassSerializer(MsgPack.Serialization.SerializationContext context) : 
                base(context) {
            this._serializer0 = context.GetSerializer<SharedModel.A.ID>(MsgPack.Serialization.EnumMessagePackSerializerHelpers.DetermineEnumSerializationMethod(context, typeof(SharedModel.A.ID), MsgPack.Serialization.EnumMemberSerializationMethod.Default));
            MsgPack.Serialization.PolymorphismSchema schema0 = default(MsgPack.Serialization.PolymorphismSchema);
            schema0 = null;
            this._serializer1 = context.GetSerializer<long>(schema0);
            MsgPack.Serialization.PolymorphismSchema schema1 = default(MsgPack.Serialization.PolymorphismSchema);
            schema1 = null;
            this._serializer2 = context.GetSerializer<string>(schema1);
            this._serializer3 = context.GetSerializer<SharedModel.A.Error>(MsgPack.Serialization.EnumMessagePackSerializerHelpers.DetermineEnumSerializationMethod(context, typeof(SharedModel.A.Error), MsgPack.Serialization.EnumMemberSerializationMethod.Default));
            this._serializer4 = context.GetSerializer<SharedModel.A2.B2.DataType>(MsgPack.Serialization.EnumMessagePackSerializerHelpers.DetermineEnumSerializationMethod(context, typeof(SharedModel.A2.B2.DataType), MsgPack.Serialization.EnumMemberSerializationMethod.Default));
            MsgPack.Serialization.PolymorphismSchema schema2 = default(MsgPack.Serialization.PolymorphismSchema);
            MsgPack.Serialization.PolymorphismSchema itemsSchema0 = default(MsgPack.Serialization.PolymorphismSchema);
            System.Collections.Generic.Dictionary<string, System.Type> itemsSchemaTypeMap0 = default(System.Collections.Generic.Dictionary<string, System.Type>);
            itemsSchemaTypeMap0 = new System.Collections.Generic.Dictionary<string, System.Type>(1);
            itemsSchemaTypeMap0.Add("0", typeof(SharedModel.A2.B2.Data1));
            itemsSchema0 = MsgPack.Serialization.PolymorphismSchema.ForPolymorphicObject(typeof(SharedModel.A2.B2.DataBase), itemsSchemaTypeMap0);
            schema2 = MsgPack.Serialization.PolymorphismSchema.ForContextSpecifiedCollection(typeof(System.Collections.Generic.List<SharedModel.A2.B2.DataBase>), itemsSchema0);
            this._serializer5 = context.GetSerializer<System.Collections.Generic.List<SharedModel.A2.B2.DataBase>>(schema2);
            MsgPack.Serialization.PolymorphismSchema schema3 = default(MsgPack.Serialization.PolymorphismSchema);
            schema3 = null;
            this._serializer6 = context.GetSerializer<System.Nullable<SharedModel.A.ID>>(schema3);
            MsgPack.Serialization.PolymorphismSchema schema4 = default(MsgPack.Serialization.PolymorphismSchema);
            schema4 = null;
            this._serializer7 = context.GetSerializer<System.Nullable<SharedModel.A.Error>>(schema4);
            MsgPack.Serialization.PolymorphismSchema schema5 = default(MsgPack.Serialization.PolymorphismSchema);
            schema5 = null;
            this._serializer8 = context.GetSerializer<System.Nullable<SharedModel.A2.B2.DataType>>(schema5);
        }
`

**Comment by yfakariya** [comment]
First, I'm not so experienced about Unity game development. But I recommend you to separate .NET code and Unity code because MsgPack will generate bit different code to avoid Unity AOT issue.
* Client (Unity) side: Generated serializers with .NET 3.5 API, and uses Unity3D or Unity3D.Full dll (use former if you use "corelib only" build settings).
* Server (.NET FW) side: Generated serializers with .NET 4.6 API, and uses .NET 4.6.1 dlls (or NuGet).

**Comment by neuecc** [comment]
I tried new 0.7.0-alpha2's Serializer generator, thank you for many improvements.

Is serializer generator supports only latest .NET or provide .NET 3.5 API mode?
I'm using generator for Unity, it does not support Task, `Func<T5...,>` and others.
I can't found any code generate options.

**Comment by yfakariya** [comment]
@neuecc Do you mean Unity does not support .NET 3.5 APIs? If not so, I guess the process generating code via .NET 3.5 mode and buiding in .NET 4.6 does not cause any problem...

**Comment by neuecc** [comment]
@yfakariya
Unity supports .NET 3.5 API, current generator(build from 0.6.8/.NET 4.6) already run on Unity.

> process generating code via .NET 3.5 mode and buiding in .NET 4.6
Yes, it is just the thing that I need.

**Comment by naotakat** [comment]
I tested serializer generated from dll built in .net 4.6 on unity. It works. iOS is on testing..
So. I wish that generating with .net 4.6 and building with unity full dll..

**Comment by naotakat** [comment]
@yfakariya @neuecc 
I solved this problem.
The solution it's like below.
I saw my stack trace. In ResolveType Method, tried to load assembly. 
A project uses Unity full dll, and B project uses .net 4.5 dll
Stack Trace let me know different public token used in each project.
public token of unity full dll is null and one of .net 4.5 is stuff generated.
So, I re-generated dll for .net 4.5 after uncheck singing the assembly.
And, It works with re-generated dll.
Finally, I can generate serializer with .net 4.5 dll and build with unity full dll.
